lib/experimental/Navigation/Carousel/DynamicCarousel/index.tsx

```diff
@@ -1,52 +1,78 @@
   const [canScrollPrev, setCanScrollPrev] = useState(false)
 
-  const scrollToNextItem = (direction: "left" | "right") => {
-    if (scrollContainerRef.current) {
-      const container = scrollContainerRef.current
-      const { scrollLeft } = container
-
-      if (direction === "right") {
-        const nextItem = Array.from(container.children).find((child) => {
-          const childLeft = (child as HTMLElement).offsetLeft
-          return childLeft - SPACE_FOR_WIDGET_SHADOW > scrollLeft
-        })
-        if (nextItem) {
-          container.scrollTo({
-            left:
-              (nextItem as HTMLElement).offsetLeft - SPACE_FOR_WIDGET_SHADOW,
-            behavior: "smooth",
-          })
-        }
-      } else {
-        const previousItem = Array.from(container.children)
-          .reverse()
-          .find((child) => {
-            const childRight =
-              (child as HTMLElement).offsetLeft +
-              (child as HTMLElement).offsetWidth
-            return childRight < scrollLeft
-          })
-        if (previousItem) {
-          container.scrollTo({
-            left:
-              (previousItem as HTMLElement).offsetLeft -
-              SPACE_FOR_WIDGET_SHADOW,
-            behavior: "smooth",
-          })
-        }
-      }
+  useLayoutEffect(() => {
+    const container = scrollContainerRef.current
+    if (!container) return
+
+    const ro = new ResizeObserver(() => updateScrollState())
+    ro.observe(container)
+
+    const handleScroll = () => {
+      updateScrollState()
+    }
+
+    container.addEventListener("scroll", handleScroll)
+    updateScrollState()
+
+    return () => {
+      ro.disconnect()
+      container.removeEventListener("scroll", handleScroll)
+    }
+  }, [])
+
+  function scrollNext() {
+    const container = scrollContainerRef.current
+    if (!container) return
+    const itemOffsetsRef = Array.from(container.children).map(
+      (child) => (child as HTMLElement).offsetLeft
+    )
+
+    const { scrollLeft } = container
+    const offsets = itemOffsetsRef
+
+    let nextOffset = offsets.find(
+      (o) => o > scrollLeft + SPACE_FOR_WIDGET_SHADOW + GAP
+    )
+    if (!nextOffset) {
+      nextOffset = container.scrollWidth
+    } else {
+      nextOffset = nextOffset - SPACE_FOR_WIDGET_SHADOW - GAP
     }
+    container.scrollTo({ left: nextOffset, behavior: "smooth" })
+  }
+
+  function scrollPrev() {
+    const container = scrollContainerRef.current
+    if (!container) return
+
+    const itemOffsetsRef = Array.from(container.children).map(
+      (child) => (child as HTMLElement).offsetLeft
+    )
+
+    const { scrollLeft } = container
+    const offsets = [...itemOffsetsRef].reverse()
+    console.log("scrollLeft", scrollLeft)
+
+    let prevOffset = offsets.find((o) => o < scrollLeft)
+    if (!prevOffset) {
+      prevOffset = 0
+    } else {
+      prevOffset = Math.max(0, prevOffset - SPACE_FOR_WIDGET_SHADOW - GAP)
+    }
```

Comment:
maybe:
```suggestion
    let prevOffset = offsets.find((o) => o < scrollLeft) || 0
    prevOffset = Math.max(0, prevOffset - SPACE_FOR_WIDGET_SHADOW - GAP)
```

I am confused why we search offset, set `prevOffset` and then set it again the next line. I suspect it should be two different variables